### PR TITLE
Add Pickscore/HPSv2 style reward training and evaluation

### PR DESCRIPTION
**Pull request by rohitrango**
# What does this PR do ?

Adds additional functionality for reward model training for text-to-image models.

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [ ] Does the trainer resume and restore model state all states?
- [ ] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [ ] Does the trainer support `max_steps=-1` and `validation`?
- [ ] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [ ] Does the trainer have proper logging?

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by rohitrango** [comment]
@terrykong @JRD971000 can someone be assigned to this PR? Thanks.

File: examples/mm/clip/conf/baseline.yaml [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    model: null
+    vocab_file: null
+    merge_file: null
+    delimiter: null # only used for tabular tokenizer
+    sentencepiece_legacy: False # Legacy=True allows you to add special tokens to sentencepiece tokenizers.
+  make_vocab_size_divisible_by: 128 # Pad the vocab size to be divisible by this value for computation efficiency.
+
+  data:
+    num_workers: 8
+    data_path: "/opt/nemo-aligner/datasets/pickapic"
```
Comment (JRD971000) [inline examples/mm/clip/conf/baseline.yaml:227]: Remove this

File: examples/mm/clip/test_reward_model.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+@torch.no_grad()
+def main(cfg) -> None:
+    logging.info("\n\n************** Experiment configuration ***********")
+    logging.info(f"\n{OmegaConf.to_yaml(cfg)}")
+
+    cfg.model.global_batch_size = cfg.trainer.devices * cfg.trainer.num_nodes * cfg.model.micro_batch_size
+
+    model = get_reward_model(cfg, cfg.model.micro_batch_size, cfg.model.global_batch_size).cuda()
+    model.eval()
+    batch_size = 250
```
Comment (JRD971000) [inline examples/mm/clip/test_reward_model.py:45]: why hard coding batch size here?

File: examples/mm/clip/test_reward_model.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    test_dl = DataLoader(
+        test_ds, batch_size=batch_size, drop_last=False, shuffle=False, collate_fn=model.dl_collate_fn
+    )
+
+    # collect all labels here
+    all_val_probs = []
+    all_val_labels = []
+
+    # run through the val and test datasets
+    thresholds = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]
```
Comment (JRD971000) [inline examples/mm/clip/test_reward_model.py:57]: What is the purpose of this?

File: examples/mm/clip/test_reward_model.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    for batch in tqdm(val_dl, total=len(val_dl)):
+        # print(batch)
```
Comment (JRD971000) [inline examples/mm/clip/test_reward_model.py:59]: remove commented lines

File: nemo_aligner/data/mm/pickscore_dataset.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+
+if __name__ == "__main__":
+    from omegaconf import OmegaConf
+    from open_clip import tokenizer
+
+    cfg = {
+        "vision": {"img_w": 224, "img_h": 224, "img_mean": OPENAI_DATASET_MEAN, "img_std": OPENAI_DATASET_STD},
+        "text": {"max_position_embeddings": 77,},
+        "data": {"data_path": "/opt/nemo-aligner/datasets/pickapic",},
```
Comment (JRD971000) [inline nemo_aligner/data/mm/pickscore_dataset.py:125]: data_path should be input to the script

File: nemo_aligner/models/mm/stable_diffusion/image_text_rms.py [hunk 6]
```diff
@@ -121,9 +122,147 @@ def model_provider_func(self, pre_process, post_process):
...
+
+    def loss_func(self, output_tensor):
+        """ 
+        rewards_0, rewards_1 - tensor of size [B,] each that computes rewards 
+        labels - Bx2 tensor containing labels
+        """
+        rewards_0, rewards_1, labels = output_tensor
+        logits = torch.stack([rewards_0, rewards_1], 1)  # [B, 2]
+        logsoftmax = F.log_softmax(logits, dim=1)
+        labels_s = labels + 1e-4
```
Comment (JRD971000) [inline nemo_aligner/models/mm/stable_diffusion/image_text_rms.py:139]: is this for stability? if so why 1e-4? The default value is usually 1e-6 in other cases; also add it to the denominator in the next line

**Review by JRD971000 — COMMENTED** [review]
@rohitrango Left a few comments

File: examples/mm/clip/test_reward_model.py (hunk 2 above)
Comment (rohitrango) [reply examples/mm/clip/test_reward_model.py:45]: no particular reason, I just found 250 to be a good batch size for iterating quickly over the entire val set. I've changed it to `cfg.model.micro_batch_size`

File: examples/mm/clip/test_reward_model.py (hunk 3 above)
Comment (rohitrango) [reply examples/mm/clip/test_reward_model.py:57]: this is how the Pickscore paper finds the optimal threshold to qualify an output as a "tie" and maximize accuracy over the validation set + report test accuracy on this parameter.

File: nemo_aligner/models/mm/stable_diffusion/image_text_rms.py (hunk 6 above)
Comment (rohitrango) [reply nemo_aligner/models/mm/stable_diffusion/image_text_rms.py:139]: 1e-4 seems to be fine if the labels are just 0, 0.5, and 1.

The denominator already has the 1e-4 added from the previous line (to normalize `label_s` to a valid pdf).

**Review by JRD971000 — DISMISSED** [review]

**Review by JRD971000 — APPROVED** [review]
